## Introduction to Data Cleaning for Real Estate Price Prediction
Accurate real estate price prediction relies heavily on the quality of the underlying data. In this project, the initial focus is on cleaning and preparing a dataset containing detailed information about flats in Gurgaon. The data cleaning process ensures that the dataset is consistent, reliable, and ready for further analysis or modeling.

### Overview of the Data Cleaning Steps
#### 1. Loading and Inspecting the Data
* The dataset is imported directly from a raw CSV file hosted online.

* Initial inspection includes viewing random samples and checking data types, which were found to be mostly object (string) types.

#### 2. Handling Unnecessary and Duplicate Data
* Columns that do not contribute to price prediction (such as link and property_id) are dropped.

* Duplicate entries are identified and can be removed to avoid redundancy.

#### 3. Managing Missing Values
* A column-wise check is performed to identify missing values.

* For columns like bedRoom, rows with missing values are removed if other critical fields are also missing.

* For categorical fields such as additionalRoom and facing, missing values are filled with a placeholder ('not available') and standardized to lowercase for consistency.

#### 4. Data Type Conversion and Feature Engineering
* Numerical columns stored as strings (e.g., price, pricePerSqft, bedRoom, bathroom) are converted to appropriate numeric types after cleaning and formatting.

* The balcony column, which sometimes contains ambiguous values like '3+', is cleaned by extracting numeric values and handling 'No' as zero.

* The floorNum column is standardized by converting textual representations (e.g., 'Ground', 'Basement') to numeric values.

#### 5. Cleaning and Standardizing Categorical Data
* The society column, which contains many unique values and ratings, is cleaned by removing rating symbols and converting all entries to lowercase.

* Categorical columns are prepared for future encoding, with consideration given to the high cardinality of some fields.

#### 6. Feature Creation
* New features such as area (calculated from price and price per square foot) and property_type are created to enrich the dataset and provide more predictive power.

#### 7. Exporting the Cleaned Data
* The final cleaned dataset is saved as a new CSV file, ready for use in modeling and analysis.

In [1]:
import numpy as np
import pandas as pd

In [2]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [3]:
base_url = "https://raw.githubusercontent.com/pranta-iitp/Real-Estate-Property-Price-Prediction-Project/main/flats.csv"
df = pd.read_csv(base_url)

In [4]:
df.sample(5)

property_name  \
682   2 BHK Flat in Sector 108 Gurgaon   
634    3 BHK Flat in Sector 61 Gurgaon   
2764   2 BHK Flat in Sector 78 Gurgaon   
2190   4 BHK Flat in Sector 82 Gurgaon   
2477  2 BHK Flat in Sector 63A Gurgaon   

                                                   link  \
682   https://www.99acres.com/2-bhk-bedroom-apartmen...   
634   https://www.99acres.com/3-bhk-bedroom-apartmen...   
2764  https://www.99acres.com/2-bhk-bedroom-apartmen...   
2190  https://www.99acres.com/4-bhk-bedroom-apartmen...   
2477  https://www.99acres.com/2-bhk-bedroom-apartmen...   

                          society       price             area  \
682   Experion The Heartsong3.9 ★      98 Lac  ₹ 13,343/sq.ft.   
634             Pioneer Park3.8 ★  2.04 Crore  ₹ 14,623/sq.ft.   
2764         Umang Monsoon Breeze      70 Lac   ₹ 5,719/sq.ft.   
2190       Mapsko Casa Bella3.8 ★  1.75 Crore   ₹ 6,903/sq.ft.   
2477    Signature Global City 63A  1.65 Crore  ₹ 15,263/sq.ft.   

                                           areaWithType     bedRoom  \
682   Super Built up area 1283(119.19 sq.m.)Built Up...  2 Bedrooms   
634   Super Built up area 1795(166.76 sq.m.)Built Up...  3 Bedrooms   
2764                 Built Up area: 1224 (113.71 sq.m.)  2 Bedrooms   
2190  Built Up area: 2535 (235.51 sq.m.)Carpet area:...  4 Bedrooms   
2477             Super Built up area 1081(100.43 sq.m.)  2 Bedrooms   

         bathroom       balcony additionalRoom  \
682   3 Bathrooms   3 Balconies     Study Room   
634   3 Bathrooms  3+ Balconies   Servant Room   
2764  2 Bathrooms   2 Balconies            NaN   
2190  5 Bathrooms  3+ Balconies            NaN   
2477  2 Bathrooms   2 Balconies            NaN   

                                                address             floorNum  \
682                Sector 108 Gurgaon, Gurgaon, Haryana   4th   of 14 Floors   
634   Near To Sector 56 Rapid Metro Station Gurgaon,...  16th   of 32 Floors   
2764  Sector 78,gurgaon, Sector 78 Gurgaon, Gurgaon,...   5th   of 12 Floors   
2190     Sector 82, Sector 82 Gurgaon, Gurgaon, Haryana    1st   of 1 Floors   
2477               Sector 63A Gurgaon, Gurgaon, Haryana    1st   of 4 Floors   

     facing     agePossession  \
682    East   1 to 5 Year Old   
634   South  5 to 10 Year Old   
2764    NaN         undefined   
2190   West         undefined   
2477    NaN          Oct 2024   

                                        nearbyLocations  \
682   ['Galleria 108 Mall', 'Dwarka Expressway', 'Ce...   
634   ['Sector 55-56 Rapid Metro', 'Hong Kong Bazaar...   
2764  ['Proposed Metro Station', 'Mahapal Shing', 'N...   
2190  ['Vatika City Centre Mall', 'Pataudi Road', 'B...   
2477  ['Sector 54 Chowk Metro Station', 'Paras Trini...   

                                            description  \
682   Corner and airy flat with ample of sunlight th...   
634   An excellent 3 bhk residential apartment is fo...   
2764  2bhk multistorey apartment for resale in umang...   
2190  Low floor easy for access , well maintained lo...   
2477  It's a 2bhk in a popular area (Adjoining humon...   

                                         furnishDetails  \
682   ['1 Bed', '1 Sofa', '8 Light', '3 AC', '1 Curt...   
634   ['3 Wardrobe', '11 Fan', '1 Exhaust Fan', '4 G...   
2764  ['1 Light', 'No AC', 'No Bed', 'No Chimney', '...   
2190  ['1 Light', 'No AC', 'No Bed', 'No Chimney', '...   
2477  ['1 Wardrobe', '1 Fan', '1 Light', '1 AC', 'No...   

                                               features  \
682   ['Security / Fire Alarm', 'Feng Shui / Vaastu ...   
634   ['Centrally Air Conditioned', 'Water purifier'...   
2764                                                NaN   
2190                                                NaN   
2477                                                NaN   

                                                 rating property_id  
682   ['Green Area5 out of 5', 'Construction4 out of...   P69721330  
634   ['Green Area5 out of 5', 'Construction3 out of

#### Data types of all the columns is object(string) Which is needed to be changed

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3017 entries, 0 to 3016
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    3017 non-null   object
 1   link             3017 non-null   object
 2   society          3016 non-null   object
 3   price            3007 non-null   object
 4   area             3004 non-null   object
 5   areaWithType     3008 non-null   object
 6   bedRoom          3008 non-null   object
 7   bathroom         3008 non-null   object
 8   balcony          3008 non-null   object
 9   additionalRoom   1694 non-null   object
 10  address          3002 non-null   object
 11  floorNum         3006 non-null   object
 12  facing           2127 non-null   object
 13  agePossession    3007 non-null   object
 14  nearbyLocations  2913 non-null   object
 15  description      3008 non-null   object
 16  furnishDetails   2203 non-null   object
 17  features         2594 non-null   

In [6]:
df.duplicated().sum()

np.int64(0)

#### Column wise missing values

In [7]:
df.isnull().sum()

property_name         0
link                  0
society               1
price                10
area                 13
areaWithType          9
bedRoom               9
bathroom              9
balcony               9
additionalRoom     1323
address              15
floorNum             11
facing              890
agePossession        10
nearbyLocations     104
description           9
furnishDetails      814
features            423
rating              341
property_id           9
dtype: int64

In [8]:
df.columns

Index(['property_name', 'link', 'society', 'price', 'area', 'areaWithType',
       'bedRoom', 'bathroom', 'balcony', 'additionalRoom', 'address',
       'floorNum', 'facing', 'agePossession', 'nearbyLocations', 'description',
       'furnishDetails', 'features', 'rating', 'property_id'],
      dtype='object')

#### Two Columns(link,property_id) will not be helpful. So we permanently drop those columns.

In [9]:
df.drop(columns=['link','property_id'],inplace=True)

In [10]:
df.columns

Index(['property_name', 'society', 'price', 'area', 'areaWithType', 'bedRoom',
       'bathroom', 'balcony', 'additionalRoom', 'address', 'floorNum',
       'facing', 'agePossession', 'nearbyLocations', 'description',
       'furnishDetails', 'features', 'rating'],
      dtype='object')

In [11]:
df.shape

(3017, 18)

In [12]:
df.rename(columns={'area':'pricePerSqft'},inplace=True)

In [13]:
df.columns

Index(['property_name', 'society', 'price', 'pricePerSqft', 'areaWithType',
       'bedRoom', 'bathroom', 'balcony', 'additionalRoom', 'address',
       'floorNum', 'facing', 'agePossession', 'nearbyLocations', 'description',
       'furnishDetails', 'features', 'rating'],
      dtype='object')

In [14]:
df['society'].value_counts()

society
SS The Leaf3.8 ★                                     73
Tulip Violet4.3 ★                                    40
Shapoorji Pallonji Joyville Gurugram4.0 ★            39
Signature Global Park4.0 ★                           36
Shree Vardhman Victoria3.8 ★                         35
Tulip Violet4.2 ★                                    33
Smart World Gems                                     32
Emaar MGF Emerald Floors Premier3.8 ★                32
Smart World Orchard                                  32
DLF The Ultima4.0 ★                                  31
Paras Dews                                           31
DLF Regal Gardens3.9 ★                               30
Shree Vardhman Flora3.8 ★                            29
M3M Woodshire4.0 ★                                   29
La Vida by Tata Housing                              28
Signature Global Solera3.7 ★                         28
Godrej Nature Plus                                   27
Emaar Gurgaon Greens4.1 ★                            25
BPTP Terra3.8 ★                                      25
Experion The Heartsong3.9 ★                          24
Eldeco Accolade3.8 ★                                 24
Vatika Gurgaon 213.7 ★                               24
DLF New Town Heights 13.9 ★                          24
Bestech Park View Residency3.9 ★                     23
M3M Heights                                          22
Tata Primanti4.1 ★                                   22
ATS Triumph3.9 ★                                     22
Emaar Palm Gardens4.1 ★                              22
ROF Ananda                                           21
Sare Crescent Parc3.5 ★                              21
Godrej 1014.0 ★                                      20
Experion Windchants3.8 ★                             20
Sare Homes                                           19
Antriksh Heights3.7 ★                                19
Ambience Creacions                                   18
ATS Tourmaline3.9 ★                                  18
Indiabulls Centrum Park3.7 ★                         18
Mapsko Mount Ville4.1 ★                              18
Supertech Araville                                   18
CHD Avenue 713.6 ★                                   18
Sobha City                                           18
M3M Golfestate4.4 ★                                  17
Conscient Heritage Max4.3 ★                          17
GLS Arawali Homes3.8 ★                               17
Godrej Summit3.7 ★                                   17
Pareena Coban Residences                             17
M3M Skywalk                                          17
Bestech Park View Grand Spa4.3 ★                     16
M3M Merlin3.7 ★                                      16
HCBS Sports Ville3.6 ★                               16
Umang Monsoon Breeze                                 15
Emaar MGF The Palm Drive4.1 ★                        15
Ansal Heights 86                                     15
Godrej Oasis3.9 ★                                    15
Umang Winter Hills3.9 ★                              14
Ramsons Kshitij3.7 ★                                 14
DLF The Primus3.9 ★                                  14
Orchid Petals4.1 ★                                   14
The Close North4.0 ★                                 14
Ireo Skyon4.0 ★                                      14
Vatika The Seven Lamps3.8 ★                          14
Ireo Victory Valley4.1 ★                             14
Indiabulls Enigma                                    14
Corona Optus3.9 ★                                    14
Imperia The Esfera                                   14
MVN Athens3.7 ★                                      14
Ireo The Corridors                                   13
Bestech Park View Ananda4.2 ★                        13
Puri Emerald Bay3.7 ★                                13
Central Park Flower Valley                           13
M3M Soulitude                                        11
DLF The Skycourt

In [15]:
df['society'].nunique()

638

#### The column *'society'* is a categorical column and it has 638 unique values. So now if we apply one hot encoding the number of columns will increase drastically.

In [16]:
# Convert each value of society into lower case and remove the rating part
import re
df['society'] = df['society'].apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip()).str.lower()

In [17]:
df['society'].nunique()

604

In [18]:
#price
df['price'].nunique()

439

In [19]:
# Check if all values in the specified column are numerical
is_numerical = pd.to_numeric(df['price'], errors='coerce').notna().all()

In [20]:
is_numerical

np.False_

In [21]:
df['price'].unique()

array(['45 Lac', '50 Lac', '40 Lac', '1.47 Crore', '70 Lac', '41 Lac',
       '2 Crore', '1.8 Crore', '1.1 Crore', '4.75 Crore', '96 Lac',
       '29 Lac', '1.35 Crore', '95 Lac', '3.95 Crore', '90 Lac',
       '1.05 Crore', nan, '2.2 Crore', '1.01 Crore', '1.85 Crore',
       '86 Lac', '2.85 Crore', 'Price on Request', '42 Lac', '6.15 Crore',
       '6.25 Crore', '1.6 Crore', '3.25 Crore', '85 Lac', '75 Lac',
       '82 Lac', '29.99 Lac', '78 Lac', '74 Lac', '3.2 Crore',
       '1.3 Crore', '25 Lac', '1.99 Crore', '1.83 Crore', '2.25 Crore',
       '2.8 Crore', '83 Lac', '80 Lac', '1.25 Crore', '23 Lac', '30 Lac',
       '1.55 Crore', '79 Lac', '99 Lac', '1.9 Crore', '1 Crore',
       '2.5 Crore', '55 Lac', '65 Lac', '31.75 Lac', '93 Lac',
       '1.2 Crore', '56 Lac', '2.7 Crore', '1.45 Crore', '46 Lac',
       '4.5 Crore', '64 Lac', '28 Lac', '3.87 Crore', '1.38 Crore',
       '43 Lac', '28.5 Lac', '1.75 Crore', '2.1 Crore', '1.29 Crore',
       '1.65 Crore', '24 Lac', '31.5 Lac', '

In [22]:
df[df['price'] == 'Price on Request'].shape

(11, 18)

In [23]:
df = df[df['price'] != 'Price on Request']

In [24]:
# convert all prices to crore
def treat_price(x):
    if type(x) == float:
        return x
    else:
        if x[1] == 'Lac':
            return round(float(x[0])/100,2)
        else:
            return round(float(x[0]),2)

In [25]:
df['price'] = df['price'].str.split(' ').apply(treat_price).astype('float')

/tmp/ipython-input-25-3314369693.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price'] = df['price'].str.split(' ').apply(treat_price).astype('float')


In [26]:
df['price'].unique()

array([ 0.45,  0.5 ,  0.4 ,  1.47,  0.7 ,  0.41,  2.  ,  1.8 ,  1.1 ,
        4.75,  0.96,  0.29,  1.35,  0.95,  3.95,  0.9 ,  1.05,   nan,
        2.2 ,  1.01,  1.85,  0.86,  2.85,  0.42,  6.15,  6.25,  1.6 ,
        3.25,  0.85,  0.75,  0.82,  0.3 ,  0.78,  0.74,  3.2 ,  1.3 ,
        0.25,  1.99,  1.83,  2.25,  2.8 ,  0.83,  0.8 ,  1.25,  0.23,
        1.55,  0.79,  0.99,  1.9 ,  1.  ,  2.5 ,  0.55,  0.65,  0.32,
        0.93,  1.2 ,  0.56,  2.7 ,  1.45,  0.46,  4.5 ,  0.64,  0.28,
        3.87,  1.38,  0.43,  1.75,  2.1 ,  1.29,  1.65,  0.24,  1.7 ,
        2.75,  0.34,  1.5 ,  0.26,  2.91,  1.15,  1.95,  0.88,  3.35,
        1.08,  2.15,  1.79,  0.22,  0.36,  2.9 ,  0.6 ,  2.4 ,  0.38,
        0.39,  0.27,  2.65,  1.32,  0.59,  1.4 ,  3.  ,  3.67,  3.85,
        3.29,  2.33,  2.79,  3.4 ,  0.72,  0.71,  1.41,  2.05,  0.54,
        1.42,  2.29,  0.97,  8.25,  4.25,  4.1 ,  0.92,  0.91,  0.66,
        0.53,  1.71,  5.7 ,  4.  ,  1.93,  3.45,  1.48,  1.49,  3.49,
        2.3 ,  2.12,

In [27]:
df.sample(5)

property_name               society  price  \
164   2 BHK Flat in Sector 95A Gurgaon  signature signum 95a   0.40   
339    2 BHK Flat in Sector 79 Gurgaon                godrej   1.29   
2676   2 BHK Flat in Sector 65 Gurgaon           m3m heights   1.90   
783    3 BHK Flat in Sector 65 Gurgaon   emaar emerald hills   1.95   
1422  2 BHK Flat in Sector 37C Gurgaon        apex our homes   0.33   

         pricePerSqft                                       areaWithType  \
164    ₹ 7,782/sq.ft.                     Carpet area: 514 (47.75 sq.m.)   
339    ₹ 8,206/sq.ft.             Super Built up area 1572(146.04 sq.m.)   
2676  ₹ 13,991/sq.ft.                 Built Up area: 1358 (126.16 sq.m.)   
783   ₹ 13,928/sq.ft.                   Carpet area: 1400 (130.06 sq.m.)   
1422   ₹ 3,567/sq.ft.  Super Built up area 925(85.94 sq.m.)Built Up a...   

         bedRoom     bathroom       balcony additionalRoom  \
164   2 Bedrooms  2 Bathrooms   2 Balconies            NaN   
339   2 Bedrooms  2 Bathrooms   3 Balconies            NaN   
2676  2 Bedrooms  2 Bathrooms   2 Balconies            NaN   
783   3 Bedrooms  3 Bathrooms  3+ Balconies     Store Room   
1422  2 Bedrooms  2 Bathrooms     1 Balcony            NaN   

                                               address             floorNum  \
164               Sector 95A Gurgaon, Gurgaon, Haryana   3rd   of 19 Floors   
339                Sector 79 Gurgaon, Gurgaon, Haryana   4th   of 15 Floors   
2676    Sector 65, Sector 65 Gurgaon, Gurgaon, Haryana  40th   of 40 Floors   
783   Amber Block, Sector 65 Gurgaon, Gurgaon, Haryana    1st   of 2 Floors   
1422         358, Sector 37C Gurgaon, Gurgaon, Haryana   3rd   of 11 Floors   

          facing    agePossession  \
164         East  0 to 1 Year Old   
339          NaN  0 to 1 Year Old   
2676         NaN        undefined   
783   North-East  0 to 1 Year Old   
1422        West  1 to 5 Year Old   

                                        nearbyLocations  \
164                                                 NaN   
339   ['Vatika Town Square-INXT', 'Naurangpur Road',...   
2676  ['Rapid Metro Sector 56', 'M3m 65th Avenue Mal...   
783   ['Emerald Plaza Shopping Mall', 'Southern Peri...   
1422  ['Elvedor Mall', 'MDS Public School', 'K.D. Ho...   

                                            description  \
164   Prime location\nAll necessary shops are in pre...   
339   This is your chance to own a 2 bhk residential...   
2676  Best in class property available at sector 65 ...   
783   Situated in sector 65 gurgaon, emaar emerald h...   
1422  This 2 bhk apartment is available for sale in ...   

                                         furnishDetails  \
164   ['3 Fan', '1 Geyser', '5 Light', '1 Chimney', ...   
339                                                 NaN   
2676  ['1 Light', 'No AC', 'No Bed', 'No Chimney', '...   
783   ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
1422  ['2 Wardrobe', '4 Fan', '1 Exhaust Fan', '2 Ge...   

                                               features  \
164   ['Security / Fire Alarm', 'Intercom Facility',...   
339                                                 NaN   
2676                                                NaN   
783   ['Security / Fire Alarm', 'Power Back-up', 'Fe...   
1422  ['Centrally Air Conditioned', 'Water purifier'...   

                                                 rating  
164   ['Environment5 out of 5', 'Safety5 out of 5', ...  
339   ['Green Area5 out of 5', 'Construction5 out of...  
2676  ['Environment4 out of 5', 'Safety4 out of 5', ...  
783   ['Green Area5 out of 5', 'Construction5 out of...  
1422  ['Environment5 out of 5', 'Lifestyle4 out of 5...

In [28]:
# pricePerSqft
df['pricePerSqft'].nunique()

2130

In [29]:
df['pricePerSqft'] = df['pricePerSqft'].str.split('/').str.get(0).str.replace('₹','').str.replace(',','').str.strip().astype('float')

In [30]:
df.sample(5)

property_name  \
1642     3 BHK Flat in Sector 72 Gurgaon   
1869     2 BHK Flat in Sector 66 Gurgaon   
123      3 BHK Flat in Sector 89 Gurgaon   
1704    3 BHK Flat in Sector 102 Gurgaon   
497   3 BHK Flat in Sector 1 Imt Manesar   

                                   society  price  pricePerSqft  \
1642                         tata primanti   3.40       15560.0   
1869                emaar mgf palm studios   1.45       14795.0   
123                          m3m soulitude   1.25        8784.0   
1704  shapoorji pallonji joyville gurugram   1.75        9449.0   
497            sbr minda sec-1 imt manesar   0.64        4193.0   

                                           areaWithType     bedRoom  \
1642             Super Built up area 2185(202.99 sq.m.)  3 Bedrooms   
1869  Super Built up area 1188(110.37 sq.m.)Carpet a...  2 Bedrooms   
123               Super Built up area 1423(132.2 sq.m.)  3 Bedrooms   
1704  Super Built up area 1852(172.06 sq.m.)Carpet a...  3 Bedrooms   
497                    Carpet area: 1526 (141.77 sq.m.)  3 Bedrooms   

         bathroom       balcony                             additionalRoom  \
1642  4 Bathrooms  3+ Balconies                               Servant Room   
1869  2 Bathrooms   3 Balconies                                        NaN   
123   3 Bathrooms   3 Balconies                          Study Room,Others   
1704  3 Bathrooms   3 Balconies  Pooja Room,Study Room,Servant Room,Others   
497   2 Bathrooms  3+ Balconies                                        NaN   

                                          address             floorNum facing  \
1642          Sector 72 Gurgaon, Gurgaon, Haryana   9th   of 40 Floors    NaN   
1869     402, Sector 66 Gurgaon, Gurgaon, Haryana   3rd   of 20 Floors   East   
123   S-68/3, Sector 89 Gurgaon, Gurgaon, Haryana    3rd   of 4 Floors   East   
1704         Sector 102 Gurgaon, Gurgaon, Haryana  10th   of 26 Floors   West   
497        Sector 1 Imt Manesar, Gurgaon, Haryana   6th   of 10 Floors   East   

         agePossession                                    nearbyLocations  \
1642  5 to 10 Year Old  ['Sector 55-56 Metro Station', 'Omaxe City Cen...   
1869   1 to 5 Year Old  ['Sri Radhe Krishna Temple', 'Icici bank ATM',...   
123    Within 6 months  ['Vatika Town Square-INXT', 'Sector 86 Road', ...   
1704   1 to 5 Year Old  ['Khan Market', 'The Esplanade Mall', 'Dwarka ...   
497       10+ Year Old  ['Orris Community Center', 'Indian Oil', 'Essa...   

                                            description  \
1642  4 bedroom 2185 sq.Ft. Middle floor apartment a...   
1869  Emaar mgf the palm drive in sector 66, gurgaon...   
123   This lovely 3 bhk apartment/flat in sector 89 ...   
1704  Looking for a 3 bhk property for sale in gurga...   
497   Flat for sale sec-1 imt manesar good property ...   

                                         furnishDetails  \
1642                                                 []   
1869  ['2 Wardrobe', '4 Fan', '3 Geyser', '5 Light',...   
123                                                 NaN   
1704  ['14 Light', '1 Modular Kitchen', '1 Chimney',...   
497                                                 NaN   

                                               features  \
1642  ['Power Back-up', 'Intercom Facility', 'Lift(s...   
1869  ['Feng Shui / Vaastu Compliant', 'Intercom Fac...   
123                                                 NaN   
1704  ['Feng Shui / Vaastu Compliant', 'Security / F...   
497   ['Power Back-up', 'Intercom Facility', 'Lift(s...   

                                                 rating  
1642  ['Green Area4 out of 5', 'Construction4 out of...  
1869  ['Environment3 out of 5', 'Lifestyle4 out of 5...  
123   ['Environment4 out of 5', 'Safety4 out of 5', ...  
1704  ['Green Area4 out of 5', 'Construction5 out of...  
497   ['Environment5 out of 5', 'Lifestyle4.5 out of...

In [31]:
# bedrooms
df[df['bedRoom'].isnull() == True]

property_name                       society  \
2849           2 BHK Flat in Sector 107 Gurgaon       signature global solera   
2850            4 BHK Flat in Sector 53 Gurgaon                tulip monsella   
2851              2 BHK Flat in New Palam Vihar                       my home   
2852                        2 BHK Flat in Sohna        breez global hill view   
2922           3 BHK Flat in Sector 99A Gurgaon      pareena coban residences   
2923         1 BHK Flat in Golf Course Ext Road  ikon tower baani city centre   
2924            2 BHK Flat in Sector 89 Gurgaon                   greenopolis   
2926  4 BHK Flat in C Block Sushant Lok Phase 1                 maple heights   
2927            2 BHK Flat in Sector 99 Gurgaon                assotech blith   

      price  pricePerSqft areaWithType bedRoom bathroom balcony  \
2849    NaN           NaN          NaN     NaN      NaN     NaN   
2850    NaN       33198.0          NaN     NaN      NaN     NaN   
2851    NaN        4400.0          NaN     NaN      NaN     NaN   
2852    NaN        5470.0          NaN     NaN      NaN     NaN   
2922    NaN        5759.0          NaN     NaN      NaN     NaN   
2923    NaN       12437.0          NaN     NaN      NaN     NaN   
2924    NaN        4820.0          NaN     NaN      NaN     NaN   
2926    NaN        8888.0          NaN     NaN      NaN     NaN   
2927    NaN        6593.0          NaN     NaN      NaN     NaN   

     additionalRoom address floorNum facing agePossession nearbyLocations  \
2849            NaN     NaN      NaN    NaN           NaN             NaN   
2850            NaN     NaN      NaN    NaN           NaN             NaN   
2851            NaN     NaN      NaN    NaN           NaN             NaN   
2852            NaN     NaN      NaN    NaN           NaN             NaN   
2922            NaN     NaN      NaN    NaN           NaN             NaN   
2923            NaN     NaN      NaN    NaN           NaN             NaN   
2924            NaN     NaN      NaN    NaN           NaN             NaN   
2926            NaN     NaN      NaN    NaN           NaN             NaN   
2927            NaN     NaN      NaN    NaN           NaN             NaN   

     description furnishDetails features rating  
2849         NaN            NaN      NaN    NaN  
2850         NaN            NaN      NaN    NaN  
2851         NaN            NaN      NaN    NaN  
2852         NaN            NaN      NaN    NaN  
2922         NaN            NaN      NaN    NaN  
2923         NaN            NaN      NaN    NaN  
2924         NaN            NaN      NaN    NaN  
2926         NaN            NaN      NaN    NaN  
2927         NaN            NaN      NaN    NaN

#### At *bedroom* column there are 9 missing values and corresponding values in other columns are also missing. That's why rows containing  missing values are removed

In [32]:
df = df[df['bedRoom'].isnull() == False]

In [33]:
df.shape

(2997, 18)

In [34]:
df['bedRoom'] = df['bedRoom'].str.split(' ').str.get(0).astype('int')

/tmp/ipython-input-34-1468100853.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['bedRoom'] = df['bedRoom'].str.split(' ').str.get(0).astype('int')


In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3016
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   society          2997 non-null   object 
 2   price            2996 non-null   float64
 3   pricePerSqft     2996 non-null   float64
 4   areaWithType     2997 non-null   object 
 5   bedRoom          2997 non-null   int64  
 6   bathroom         2997 non-null   object 
 7   balcony          2997 non-null   object 
 8   additionalRoom   1692 non-null   object 
 9   address          2991 non-null   object 
 10  floorNum         2995 non-null   object 
 11  facing           2123 non-null   object 
 12  agePossession    2996 non-null   object 
 13  nearbyLocations  2906 non-null   object 
 14  description      2997 non-null   object 
 15  furnishDetails   2200 non-null   object 
 16  features         2590 non-null   object 
 17  rating           26

In [36]:
df.sample(5)

property_name                   society  price  \
2925   1 BHK Flat in Sector 95 Gurgaon                rof ananda   0.26   
2600  3 BHK Flat in Sector 109 Gurgaon            ats tourmaline   2.30   
395   3 BHK Flat in Sector 102 Gurgaon      emaar gurgaon greens   1.35   
477    3 BHK Flat in Sector 66 Gurgaon  emaar mgf the palm drive   2.99   
2983   3 BHK Flat in Sector 95 Gurgaon   sidhartha ncr one phase   0.90   

      pricePerSqft                                       areaWithType  \
2925        6032.0  Super Built up area 431(40.04 sq.m.)Carpet are...   
2600        8897.0             Super Built up area 2585(240.15 sq.m.)   
395         8181.0  Super Built up area 1650(153.29 sq.m.)Carpet a...   
477        14950.0  Super Built up area 2000(185.81 sq.m.)Built Up...   
2983        4500.0                 Built Up area: 2000 (185.81 sq.m.)   

      bedRoom     bathroom       balcony       additionalRoom  \
2925        1   1 Bathroom     1 Balcony                  NaN   
2600        3  4 Bathrooms  3+ Balconies  Servant Room,Others   
395         3  3 Bathrooms   3 Balconies         Servant Room   
477         3  3 Bathrooms   3 Balconies         Servant Room   
2983        3  3 Bathrooms     1 Balcony                  NaN   

                                                address             floorNum  \
2925         F 801, Sector 95 Gurgaon, Gurgaon, Haryana   8th   of 14 Floors   
2600  Flat Number 2101, 10th Floor, Tower 2, Sector ...  10th   of 27 Floors   
395                Sector 102 Gurgaon, Gurgaon, Haryana   9th   of 14 Floors   
477                 Sector 66 Gurgaon, Gurgaon, Haryana   7th   of 17 Floors   
2983                Sector 95 Gurgaon, Gurgaon, Haryana  15th   of 16 Floors   

          facing    agePossession  \
2925       North  0 to 1 Year Old   
2600        East  0 to 1 Year Old   
395   North-East  1 to 5 Year Old   
477         East  1 to 5 Year Old   
2983         NaN  1 to 5 Year Old   

                                        nearbyLocations  \
2925  ['Metro', 'Dwarka Expressway', 'Rajeev Chowk',...   
2600  ['Dwarka Sector 21 Metro Station', 'NeoSquare ...   
395   ['JMS Marine Square Mall', 'Dwarka Expressway'...   
477   ['Sector 55-56 Rapid Metro Station', 'HUB 66',...   
2983  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                            description  \
2925  All amenities in societies can be seen from go...   
2600  Residential apartment for sell.Located on 10th...   
395   3 bhi plus servant,spacious flat is ready to m...   
477                   We deal exclusively in palm drive   
2983  Residential apartment for sell.Located on 15th...   

                                         furnishDetails  \
2925  ['3 Fan', '1 Exhaust Fan', '5 Light', 'No AC',...   
2600                                                 []   
395   ['3 Fan', '4 Light', '4 AC', '1 Modular Kitche...   
477   ['5 Fan', '1 Exhaust Fan', '3 Geyser', '1 Stov...   
2983  ['1 Wardrobe', '1 Fan', '1 Light', 'No AC', 'N...   

                                               features  \
2925  ['Feng Shui / Vaastu Compliant', 'Security / F...   
2600  ['Feng Shui / Vaastu Compliant', 'Security / F...   
395   ['Intercom Facility', 'Lift(s)', 'Maintenance ...   
477   ['Feng Shui / Vaastu Compliant', 'Security / F...   
2983  ['Power Back-up', 'Lift(s)', 'Swimming Pool', ...   

                                                 rating  
2925  ['Safety4.5 out of 5', 'Lifestyle4.5 out of 5'...  
2600  ['Green Area4.5 out of 5', 'Amenities4.5 out o...  
395   ['Green Area5 out of 5', 'Construction4 out of...  
477   ['Green Area5 out of 5', 'Construction4.5 out ...  
2983  ['Management4 out of 5', 'Green Area4 out of 5...

In [37]:
#bathroom
df[df['bathroom'].isnull() == True]

Empty DataFrame
Columns: [property_name, society, price, pricePerSqft, areaWithType, bedRoom, bathroom, balcony, additionalRoom, address, floorNum, facing, agePossession, nearbyLocations, description, furnishDetails, features, rating]
Index: []

#### The column 'bathroom' contains no missing value.

In [38]:
df['bathroom'] = df['bathroom'].str.split(' ').str.get(0).astype('int')

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3016
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   society          2997 non-null   object 
 2   price            2996 non-null   float64
 3   pricePerSqft     2996 non-null   float64
 4   areaWithType     2997 non-null   object 
 5   bedRoom          2997 non-null   int64  
 6   bathroom         2997 non-null   int64  
 7   balcony          2997 non-null   object 
 8   additionalRoom   1692 non-null   object 
 9   address          2991 non-null   object 
 10  floorNum         2995 non-null   object 
 11  facing           2123 non-null   object 
 12  agePossession    2996 non-null   object 
 13  nearbyLocations  2906 non-null   object 
 14  description      2997 non-null   object 
 15  furnishDetails   2200 non-null   object 
 16  features         2590 non-null   object 
 17  rating           26

In [40]:
#balcony
df[df['balcony'].isnull() == True]

Empty DataFrame
Columns: [property_name, society, price, pricePerSqft, areaWithType, bedRoom, bathroom, balcony, additionalRoom, address, floorNum, facing, agePossession, nearbyLocations, description, furnishDetails, features, rating]
Index: []

In [41]:
df.sample(3)

property_name               society  price  \
3000   3 BHK Flat in Sector 78 Gurgaon  umang monsoon breeze   1.20   
2714  4 BHK Flat in Valley View Estate    valley view estate   1.50   
1340   3 BHK Flat in Sector 92 Gurgaon            sare homes   0.68   

      pricePerSqft                                       areaWithType  \
3000        6493.0  Super Built up area 2250(209.03 sq.m.)Carpet a...   
2714        7489.0  Super Built up area 2000(185.81 sq.m.)Built Up...   
1340        5238.0                   Carpet area: 1298 (120.59 sq.m.)   

      bedRoom  bathroom       balcony additionalRoom  \
3000        3         3   2 Balconies     Pooja Room   
2714        4         4  3+ Balconies            NaN   
1340        3         3   2 Balconies     Pooja Room   

                                             address             floorNum  \
3000             Sector 78 Gurgaon, Gurgaon, Haryana   8th   of 12 Floors   
2714  Tower 14, Valley View Estate, Gurgaon, Haryana  10th   of 18 Floors   
1340       1102, Sector 92 Gurgaon, Gurgaon, Haryana  11st   of 21 Floors   

          facing     agePossession  \
3000         NaN  5 to 10 Year Old   
2714  North-East  5 to 10 Year Old   
1340  North-West   1 to 5 Year Old   

                                        nearbyLocations  \
3000  ['Proposed Metro Station', 'Mahapal Shing', 'N...   
2714  ['HANUMAN MANDIR', 'SHIV MANDIR BALIYAWAS', 'I...   
1340  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                            description  \
3000  A 3 bedroom flat, located in sector-78 gurgaon...   
2714  Awesome flat with all fittings 3 side open cor...   
1340  This beautiful 3 bhk flat in sector 92 gurgaon...   

                                         furnishDetails  \
3000  ['1 Water Purifier', '1 Fridge', '5 Fan', '1 E...   
2714                                                 []   
1340                                                NaN   

                                               features  \
3000                     ['Piped-gas', 'Natural Light']   
2714  ['Feng Shui / Vaastu Compliant', 'Lift(s)', 'M...   
1340  ['Security / Fire Alarm', 'Feng Shui / Vaastu ...   

                                                 rating  
3000  ['Safety4 out of 5', 'Lifestyle4 out of 5', 'E...  
2714  ['Environment5 out of 5', 'Safety5 out of 5', ...  
1340  ['Environment5 out of 5', 'Lifestyle4 out of 5...

#### The column *balcony* Contains value Like '3+'. So you do not know exact amount of number of balconies. That's why we are not changing its data type to integer. We are just replacing no with 0

In [42]:
df['balcony'] = df['balcony'].str.split(' ').str.get(0).str.replace('No','0')

In [43]:
# additionalRoom
df[df['additionalRoom'].isnull() == True].shape

(1305, 18)

In [44]:
df['additionalRoom'].nunique()

49

In [45]:
df['additionalRoom'].fillna('not available',inplace=True)
df['additionalRoom'] = df['additionalRoom'].str.lower()

/tmp/ipython-input-45-586694474.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['additionalRoom'].fillna('not available',inplace=True)


In [46]:
df[df['additionalRoom'].isnull() == True].shape

(0, 18)

In [47]:
df['additionalRoom'].nunique()

50

#### For the column *additionalRoom*, We have just replaced The *Nan* value with *not available*. Moreover we converted all the values to lower case so that in case there are any duplicates it will be removed

In [48]:
# floor num
df['floorNum'].nunique()

550

In [49]:
df['floorNum'].isnull().sum()

np.int64(2)

In [50]:
df[df['floorNum'].isnull() == True]

property_name             society  price  \
181   3 BHK Flat in Dwarka Expressway Gurgaon  experion heartsong   1.08   
2766          2 BHK Flat in Sector 78 Gurgaon                 nan   0.60   

      pricePerSqft                        areaWithType  bedRoom  bathroom  \
181         6150.0  Built Up area: 1758 (163.32 sq.m.)        3         3   
2766        3692.0  Built Up area: 1625 (150.97 sq.m.)        2         2   

     balcony additionalRoom  \
181        0  not available   
2766       0  not available   

                                                address floorNum facing  \
181   604, Tower B-3, 6th Floor,Sector 108, Dwarka E...      NaN    NaN   
2766       Gurgaon, Sector 78 Gurgaon, Gurgaon, Haryana      NaN    NaN   

           agePossession nearbyLocations  \
181   Under Construction             NaN   
2766  Under Construction             NaN   

                                            description furnishDetails  \
181   A property by one of the most reputed builders...             []   
2766  The property is under construction it's by rah...             []   

     features                                             rating  
181       NaN                                                NaN  
2766      NaN  ['Safety4 out of 5', 'Lifestyle4 out of 5', 'E...

In [51]:
df['floorNum'] = df['floorNum'].str.split(' ').str.get(0).replace('Ground','0').str.replace('Basement','-1').str.replace('Lower','0').str.extract(r'(\d+)')

In [52]:
df.sample(5)

property_name                      society  price  \
1546     4 BHK Flat in Sector 81 Gurgaon  bestech park view grand spa   4.70   
1346           2 BHK Flat in Palam Vihar  bestech park view residency   1.01   
1205  3 BHK Flat in Sector 1 Imt Manesar         hsiidc sidco aravali   0.81   
2347        2 BHK Flat in Sector-5 Sohna     mvn athens sohna gurgaon   0.28   
1347           3 BHK Flat in Palam Vihar  bestech park view residency   1.51   

      pricePerSqft                                       areaWithType  \
1546        6786.0             Super Built up area 6926(643.45 sq.m.)   
1346        7137.0             Super Built up area 1415(131.46 sq.m.)   
1205        3129.0  Super Built up area 2588(240.43 sq.m.)Built Up...   
2347        5833.0                     Carpet area: 480 (44.59 sq.m.)   
1347        7890.0             Super Built up area 1920(178.37 sq.m.)   

      bedRoom  bathroom balcony additionalRoom  \
1546        4         4      3+   servant room   
1346        2         2       3  not available   
1205        3         3      3+   servant room   
2347        2         2       1  not available   
1347        3         4      3+   servant room   

                                        address floorNum      facing  \
1546  1908, Sector 81 Gurgaon, Gurgaon, Haryana       19       North   
1346              Palam Vihar, Gurgaon, Haryana        0        East   
1205     Sector 1 Imt Manesar, Gurgaon, Haryana        2  South-West   
2347           Sector-5 Sohna, Gurgaon, Haryana        2         NaN   
1347              Palam Vihar, Gurgaon, Haryana        8        West   

         agePossession                                    nearbyLocations  \
1546   1 to 5 Year Old  ['Sapphire 83 Mall', 'NH-8, IMT Manesar', 'Dwa...   
1346  5 to 10 Year Old  ['Dwarka Sector 21 Metro Station', 'HUDA Marke...   
1205  5 to 10 Year Old  ['Pooja Clinic', 'Dr. Sahil Clinic', 'Prakash ...   
2347   1 to 5 Year Old                                                NaN   
1347  5 to 10 Year Old  ['Dwarka Sector 21 Metro Station', 'HUDA Marke...   

                                            description  \
1546  Semi furnished 4bhk 6926sqft penthouse with pe...   
1346  Bestech park view residency is one of gurgaon'...   
1205  Located in the popular residential address of ...   
2347  Near to dam dama lake, lush green and 3km away...   
1347  Bestech park view residency is one of the most...   

                                         furnishDetails  \
1546  ['4 Wardrobe', '1 Exhaust Fan', '4 Geyser', '4...   
1346  ['2 Wardrobe', '2 Fan', '1 Geyser', '4 Light',...   
1205                                                NaN   
2347                                                NaN   
1347  ['3 Wardrobe', '5 Fan', '5 Light', '1 Modular ...   

                                               features  \
1546  ['Security / Fire Alarm', 'Power Back-up', 'Pr...   
1346  ['Security / Fire Alarm', 'Power Back-up', 'Li...   
1205  ['Security / Fire Alarm', 'Power Back-up', 'Fe...   
2347  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1347  ['Security / Fire Alarm', 'Power Back-up', 'Li...   

                                                 rating  
1546  ['Green Area5 out of 5', 'Construction5 out of...  
1346  ['Green Area5 out of 5', 'Construction4 out of...  
1205  ['Environment5 out of 5', 'Lifestyle4.5 out of...  
2347                                                NaN  
1347  ['Green Area5 out of 5', 'Construction4 out of...

In [53]:
df['facing'].value_counts()

facing
North-East    505
East          490
North         301
South         203
West          183
North-West    162
South-East    144
South-West    135
Name: count, dtype: int64

In [54]:
df['facing'].isnull().sum()

np.int64(874)

In [55]:
df['facing'].fillna('not available',inplace=True)
df['facing'] = df['facing'].str.lower()

/tmp/ipython-input-55-44552333.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['facing'].fillna('not available',inplace=True)


#### We are creating a new feature column *area* And *property_type*

In [56]:
df.insert(loc=4,column='area',value=round((df['price']*10000000)/df['pricePerSqft']))
df.insert(loc=1,column='property_type',value='flat')

In [57]:
df.sample(5)

property_name property_type                society  \
215   2 BHK Flat in Sector 79 Gurgaon          flat         m3m golfestate   
2298  2 BHK Flat in Sector 68 Gurgaon          flat        pareena mi casa   
1154  3 BHK Flat in Sector 61 Gurgaon          flat           pioneer park   
2411  3 BHK Flat in Sector 86 Gurgaon          flat  pyramid urban homes 2   
2882    2 BHK Flat in Dayanand Colony          flat       divya apartments   

      price  pricePerSqft    area  \
215     1.4       10000.0  1400.0   
2298    1.0        8163.0  1225.0   
1154    2.1       11666.0  1800.0   
2411    0.6        8000.0   750.0   
2882    0.6        7594.0   790.0   

                                           areaWithType  bedRoom  bathroom  \
215                    Carpet area: 1400 (130.06 sq.m.)        2         2   
2298                   Carpet area: 1225 (113.81 sq.m.)        2         2   
1154  Super Built up area 1800(167.23 sq.m.)Carpet a...        3         3   
2411                     Carpet area: 750 (69.68 sq.m.)        3         2   
2882  Built Up area: 790 (73.39 sq.m.)Carpet area: 7...        2         2   

     balcony                      additionalRoom  \
215        2  study room,servant room,store room   
2298       2                        servant room   
1154       3                       not available   
2411       2                       not available   
2882       0                       not available   

                                                address floorNum  \
215                 Sector 79 Gurgaon, Gurgaon, Haryana        4   
2298                Sector 68 Gurgaon, Gurgaon, Haryana       16   
1154                Sector 61 Gurgaon, Gurgaon, Haryana       15   
2411                Sector 86 Gurgaon, Gurgaon, Haryana        2   
2882  104,first Floor, Dayanand Colony, Gurgaon, Har...        2   

             facing    agePossession  \
215            east              NaN   
2298  not available  0 to 1 Year Old   
1154     south-east  1 to 5 Year Old   
2411           east  Within 6 months   
2882     south-west  1 to 5 Year Old   

                                        nearbyLocations  \
215   ['Petrol Pump Indian Oil', 'Petrol Pump', 'Rao...   
2298  ['Sector 55-56 Metro Station', 'Airia Mall', '...   
1154  ['Sector 55-56 Rapid Metro', 'Hong Kong Bazaar...   
2411  ['Sapphire 83 Mall', 'Rampura Flyover, Naurang...   
2882  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   

                                            description  \
215   Individual bar plus ground floor room , two si...   
2298  Inner park facing, abundant sunlight, close pr...   
1154  3bhk pioneer park sector 61 gurugram\n Additio...   
2411  Best location with first open bolcony type of ...   
2882  We are the proud owners of this 2 bhk apartmen...   

                                         furnishDetails  \
215   ['1 Water Purifier', '1 Fan', '1 Fridge', '1 D...   
2298                                                NaN   
1154  ['3 Wardrobe', '1 Water Purifier', '4 Fan', '1...   
2411                                                NaN   
2882  ['3 Wardrobe', '3 Fan', '1 Exhaust Fan', '6 Li...   

                                               features  \
215   ['Intercom Facility', 'Lift(s)', 'High Ceiling...   
2298  ['Feng Shui / Vaastu Compliant', 'Intercom Fac...   
1154  ['Feng Shui / Vaastu Compliant', 'Security / F...   
2411  ['Feng Shui / Vaastu Compliant', 'Lift(s)', 'P...   
2882  ['Lift(s)', 'Maintenance Staff', 'Recently Ren...   

                                                 rating  
215   ['Environment4 out of 5', 'Safety4 out of 5', ...  
2298  ['Environment4 out of 5', 'Lifestyle4 out of 5...  
1154  ['Green Area5 out of 5', 'Construction3 out of...  
2411  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
2882  ['Safety4 out of 5', 'Lifestyle5 out of 5', 'E...

In [58]:
df.to_csv('cleaned_flats.csv',index=False)